torch/distributed/_tensor/dispatch.py

```diff
@@ -198,11 +196,22 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
         # communicate the result to all ranks for some operators that return scalar value
         if output_sharding.output_spec is None:
             if op_call == aten.equal.default:
-                obj_list = [None for _ in range(dist.get_world_size())]
-                dist.all_gather_object(obj_list, local_results)  # type: ignore[possibly-undefined]
-                obj_list = list(filter(lambda x: x is not None, obj_list))
-                # perform reduce on the collection with AND op
-                local_results = functools.reduce(operator.and_, obj_list, True)
+                res_tensor_list = [
+                    torch.tensor(False, dtype=torch.bool, device=mesh.device_type)
+                    for _ in range(dist.get_world_size())
+                ]
+                if local_results is None:
```

Comment:
> Is it because some ranks may not have input at all.

not exactly but very close. In most cases we have full device mesh (i.e. this mesh consists of all ranks in WORLD) but in some cases we have sub-mesh (i.e. mesh only covers a subset of ranks). In sub-mesh case, if the current rank is not a part of mesh, we have `local_results` be `None`.